tsdb/head_append_v2_test.go

```diff
@@ -0,0 +1,4655 @@
+	}
+
+	// Rollback after a failed append and test if the low watermark has progressed anyway.
+	hb, _ := newTestHead(t, 1000, compression.None, false)
+	defer func() {
+		require.NoError(t, hb.Close())
+	}()
+
+	app := hb.AppenderV2(context.Background())
+	_, err := app.Append(0, labels.FromStrings("foo", "bar"), 0, 0, 0, nil, nil, storage.AOptions{})
+	require.NoError(t, err)
+	require.NoError(t, app.Commit())
+	require.Equal(t, uint64(1), hb.iso.lowWatermark())
+
+	app = hb.AppenderV2(context.Background())
+	_, err = app.Append(0, labels.FromStrings("foo", "bar"), 0, 1, 1, nil, nil, storage.AOptions{})
+	require.NoError(t, err)
+	_, err = app.Append(0, labels.FromStrings("foo", "bar", "foo", "baz"), 0, 2, 2, nil, nil, storage.AOptions{})
+	require.Error(t, err)
+	require.NoError(t, app.Rollback())
+	require.Equal(t, uint64(2), hb.iso.lowWatermark())
+
+	app = hb.AppenderV2(context.Background())
+	_, err = app.Append(0, labels.FromStrings("foo", "bar"), 0, 3, 3, nil, nil, storage.AOptions{})
+	require.NoError(t, err)
+	require.NoError(t, app.Commit())
+	require.Equal(t, uint64(3), hb.iso.lowWatermark(), "Low watermark should proceed to 3 even if append #2 was rolled back.")
+}
+
+func TestHeadAppenderV2_IsolationLowWatermarkMonotonous(t *testing.T) {
+	if defaultIsolationDisabled {
+		t.Skip("skipping test since tsdb isolation is disabled")
+	}
+
+	hb, _ := newTestHead(t, 1000, compression.None, false)
+	defer func() {
+		require.NoError(t, hb.Close())
+	}()
+
+	app1 := hb.AppenderV2(context.Background())
+	_, err := app1.Append(0, labels.FromStrings("foo", "bar"), 0, 0, 0, nil, nil, storage.AOptions{})
+	require.NoError(t, err)
+	require.NoError(t, app1.Commit())
+	require.Equal(t, uint64(1), hb.iso.lowWatermark(), "Low watermark should by 1 after 1st append.")
+
+	app1 = hb.AppenderV2(context.Background())
+	_, err = app1.Append(0, labels.FromStrings("foo", "bar"), 0, 1, 1, nil, nil, storage.AOptions{})
+	require.NoError(t, err)
+	require.Equal(t, uint64(2), hb.iso.lowWatermark(), "Low watermark should be two, even if append is not committed yet.")
+
+	app2 := hb.AppenderV2(context.Background())
+	_, err = app2.Append(0, labels.FromStrings("foo", "baz"), 0, 1, 1, nil, nil, storage.AOptions{})
+	require.NoError(t, err)
+	require.NoError(t, app2.Commit())
+	require.Equal(t, uint64(2), hb.iso.lowWatermark(), "Low watermark should stay two because app1 is not committed yet.")
+
+	is := hb.iso.State(math.MinInt64, math.MaxInt64)
+	require.Equal(t, uint64(2), hb.iso.lowWatermark(), "After simulated read (iso state retrieved), low watermark should stay at 2.")
+
+	require.NoError(t, app1.Commit())
+	require.Equal(t, uint64(2), hb.iso.lowWatermark(), "Even after app1 is committed, low watermark should stay at 2 because read is still ongoing.")
+
+	is.Close()
+	require.Equal(t, uint64(3), hb.iso.lowWatermark(), "After read has finished (iso state closed), low watermark should jump to three.")
+}
+
```

Comment:
Missing `TestHeadSeriesChunkRace` equivalent with AppenderV2

Reply:
I'd skip, not related to appending.